### DISCUSSION TITLE
input으로 파일 업로드를 하지 않는 이유 / fakePath

### DISCUSSION THREAD

**Discussion by devwontory** [KGU-CS-HOME/.github#4] [3. 이슈]
** 23.01.06 미팅에서 나누었던 input으로 파일 업로드를 시도하면 DB에 저장되지 않는 이유에 대한 내용입니다. **

발표에서 제가 조사했던 자료의 일부입니다.

> 1. 파일이 서블릿으로 오면 기본적으로 application/x-www-form-urlencoded으로 문자열 인코딩되서 String으로 Parameter를 얻어올 수 있는데 이미지 데이터는 **바이트코드**로 얻어와야 해서 form 내부의 인코딩 방식을 변경해 전달해야 한다. **enctype=”multipart/form-data”**
> 2. HttpServletRequest는 파일 업로드를 지원하지 않는다.
>    * HttpServletRequest는 웹 클라이언트가 전달하는 Multipart 데이터를 쉽게 처리하는 메소드를 제공하지 않음. → http 프로토콜의 body 부분을 읽어들이는 input 스트림만을 지원.
>    * 사용자는 input 스트림을 이용해서 Multipart 부분을 잘 나누어서 사용해야 한다. → 보통은 라이브러리를 사용 ex) Apache의 commons-fileupload
> 3. input type=”file”은 들어온 자료의 **fakePath** + 업로드된 원본 파일명이 보이게 된다. (fakePath: 브라우저가 만든 가상의 주소)
> 4. type=”file”에 입력을 한번 더 번복하면 value에 공백이 들어온다. (브라우저 보안 정책)

이 중 3번의 내용을 보충하자면, 일부 브라우저는 자바스크립트가 로컬의 파일에 접근하는 것을 막고자 fakePath라는 것을 사용하여 온전한 경로의 공개를 막는 보안 정책을 갖고 있습니다.

**Comment by noahyunjun** [comment]
이슈라고 판단하긴 어렵지만, 잡담보다는 이런 내용들을 모두 이슈탭에서 볼 수 있으면 좋을 것 같습니다.

**Reply by devwontory** [reply]
이슈로 수정하였습니다!

**Comment by gabrielyoon7** [comment]
좋은 글 이네요
